README.md
# Argon2 - Ruby Wrapper

A ruby wrapper for the Argon2 password hashing algorithm.

*This is an independent project, and not official from the PHC team.*

This gem provides a 1:1 replacement for the `argon2` gem, with various
improvements. Want to know more about why `argon2` was forked?
[Read more](#why-fork-argon2)

Wish to upgrade an existing application to use the improved API?
[Migration guide](#migrating-from-argon2-to-sorcery-argon2)

This fork is kept up-to-date with `argon2`, latest sync: `argon2 - v2.2.0`

## Table of Contents

1. [Useful Links](#useful-links)
2. [API Summary](#api-summary)
3. [Installation](#installation)
4. [Why fork `argon2`?](#why-fork-argon2)
5. [Migrating from `argon2` to `sorcery-argon2`](#migrating-from-argon2-to-sorcery-argon2)
6. [Contributing](#contributing)
7. [Contact](#contact)
8. [License](#license)

## Useful Links

* [Documentation](https://rubydoc.info/gems/sorcery-argon2)

## API Summary

Below is a summary of the library methods. Most method names are self explaining
and the rest are commented:

### Argon2::Password

```ruby
# Class methods
Argon2::Password.create(password, options = {})
Argon2::Password.valid_hash?(digest)
Argon2::Password.verify_password(password, digest, pepper = nil)

# Instance Methods
argon2 = Argon2::Password.new(digest)
argon2 == other_argon2
argon2.matches?(password, pepper = nil)
argon2.to_s   # Returns the digest as a String
argon2.to_str # Also returns the digest as a String

# Argon2::Password Attributes (readonly)
argon2.digest
argon2.variant
argon2.version
argon2.t_cost
argon2.m_cost
argon2.p_cost
argon2.salt
argon2.checksum
```

### Errors

```ruby
Argon2::Error
Argon2::Errors::InvalidHash
Argon2::Errors::InvalidVersion
Argon2::Errors::InvalidCost
Argon2::Errors::InvalidTCost
Argon2::Errors::InvalidMCost
Argon2::Errors::InvalidPCost
Argon2::Errors::InvalidPassword
Argon2::Errors::InvalidSaltSize
Argon2::Errors::InvalidOutputLength
Argon2::Errors::ExtError
```

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'sorcery-argon2'
```

And then execute:

```
$ bundle
```

Or install it yourself as:

```
$ gem install sorcery-argon2
```

Require Sorcery-Argon2 in your project:

```ruby
require 'argon2'
```

## Why fork `argon2`?

While implementing Argon2 support in Sorcery v1, I noticed that the current
ruby wrapper (`argon2` - [technion/ruby-argon2](https://github.com/technion/ruby-argon2))
had some questionable design decisions, and attempted to address them through a
pull request. The sole maintainer of the gem rejected these changes summarily,
without pointing out any specific concerns other than not understanding why the
changes were necessary. This lead to me ([@joshbuker](https://github.com/joshbuker))
being directed to create a fork instead:
[technion/ruby-argon2#44](https://github.com/technion/ruby-argon2/pull/44#issuecomment-816271661)

### Why should I trust this fork?

You shouldn't trust this code more than you trust any other open source project.
It's written by someone you don't know, and even if there is no malicious
intent, there is no guarantee that the code is secure. Open source security is
driven by having the community vett popular libraries, and discovering flaws
through the sheer number of intelligent community members looking at the code.

That being said, the original library `argon2` also falls under the same
category. Ultimately, it was also written by a single person and is not
thoroughly vetted by the community at the time of writing. A community member
([@joshbuker](https://github.com/joshbuker), in this case) finding flaws in the
implementation, and the fixes being rejected from upstream, is how this fork
came into being.

### What are the changes, why are they necessary?

The Argon2::Password interface was, to put it bluntly, poorly executed in the
original library. The Password class instance was not a representation of an
Argon2 password as one would expect, but instead an unnecessary abstraction
layer used to store the settings passed to the underlying Argon2 C Library. This
not only led to an overly complicated method of generating Argon2 hashes, but
also meant that the class could not be used to read data back out of an Argon2
digest.

Originally, to generate an Argon2 hash/digest, one would have to do the
following:

```ruby
# Create an instance of the Argon2::Password class to store your options:
instance = Argon2::Password.new(t_cost: 4, m_cost: 16)
# Use this instance to generate the hash by calling create:
instance.create(password)
    => "$argon2i$v=19$m=65536,t=2,p=1$jL7lLEAjDN+pY2cG1N8D2g$iwj1ueduCvm6B9YVjBSnAHu+6mKzqGmDW745ALR38Uo"
```

Not only is this abstraction step unnecessary, it opens up a new way for
developers to make a security mistake. New salts are only generated on the
creation of a new Argon2::Password instance, meaning if you reuse the instance,
those passwords will share the same salt.

```ruby
instance = Argon2::Password.new(t_cost: 4, m_cost: 16)
# digest1 and digest2 will share the same salt:
digest1 = instance.create(password1)
digest2 = instance.create(password2)
```

Also, because of how the instance of Argon2::Password was designed, it cannot be
used for reading information back out of an Argon2::Password. This is a summary
of the original Argon2::Password API:

```ruby
# Class methods
Argon2::Password.create(password) # Uses the default options to create a digest
Argon2::Password.valid_hash?(digest)
Argon2::Password.verify_password(password, digest, pepper = nil)

# Instance Methods
argon2 = Argon2::Password.new(options = {}) # Purely for storing options
argon2.create(password) # Take the options and generate an Argon2 digest
```

Compare this with `sorcery-argon2`:

```ruby
# Class methods
Argon2::Password.create(password, options = {}) # Same as before but accepts passing options
Argon2::Password.valid_hash?(digest)
Argon2::Password.verify_password(password, digest, pepper = nil)

# Instance Methods
argon2 = Argon2::Password.new(digest) # Now represents an Argon2 digest
argon2 == other_argon2 # Which can be compared with `==` against other Argon2::Password instances
argon2.matches?(password, pepper = nil) # Or against the original password
argon2.to_s   # Returns the digest as a String
argon2.to_str # Also returns the digest as a String

# Argon2::Password Attributes (readonly)
argon2.digest
argon2.variant
argon2.version
argon2.t_cost
argon2.m_cost
argon2.p_cost
argon2.salt
argon2.checksum
```

Another minor issue is that all library errors fall to a single non-descriptive
class:

```ruby
Argon2::ArgonHashFail
```

Compare with `sorcery-argon2`:

```ruby
Argon2::Error # Replaces `Argon2::ArgonHashFail`

# The following errors all inherit from Argon2::Error, and allow you to catch
# specifically the error you're interested in:
Argon2::Errors::InvalidHash
Argon2::Errors::InvalidVersion
Argon2::Errors::InvalidCost
Argon2::Errors::InvalidTCost
Argon2::Errors::InvalidMCost
Argon2::Errors::InvalidPCost
Argon2::Errors::InvalidPassword
Argon2::Errors::InvalidSaltSize
Argon2::Errors::InvalidOutputLength
Argon2::Errors::ExtError
```

Finally, the original library documentation is not only incomplete, but
straight up broken/inaccurate in some areas. `sorcery-argon2` has fixed these
issues, and has 100% documentation of the API.

* [`argon2` Documentation](https://rubydoc.info/gems/argon2)
* [`sorcery-argon2` Documentation](https://rubydoc.info/gems/sorcery-argon2)

## Migrating from `argon2` to `sorcery-argon2`

There are two primary changes going from `argon2` to `sorcery-argon2`:

### The Argon2::Password API has been refactored

*Argon2::Password.new and Argon2::Password.create are now different.*

Argon2::Passwords can now be created without initializing an instance first.

To upgrade:

```ruby
# Take instances where you abstract creating the password by first exposing an
# Object instance:
instance = Argon2::Password.new(m_cost: some_m_cost)
instance.create(input_password)

# And remove the abstraction step:
Argon2::Password.create(input_password, m_cost: some_m_cost)
```

*Argon2::Password.create no longer accepts custom salts.*

You should not be providing your own salt to the Argon2 algorithm (this library
does it for you). Previously you could pass an option of `salt_do_not_supply`,
which has been removed in `sorcery-argon2 - v1.0.0`.

### The errors have been restructured

**The root level error has been renamed.**

Argon2::ArgonHashFail has been renamed to Argon2::Error

To upgrade:

```ruby
# Find any instances of Argon2::ArgonHashFail, for example...
def login(username, password)
  [...]
rescue Argon2::ArgonHashFail
  [...]
end

# And do a straight 1:1 replacement
def login(username, password)
  [...]
rescue Argon2::Error
  [...]
end
```

## Contributing

Bug reports and pull requests are welcome on GitHub at
[Sorcery/argon2](https://github.com/Sorcery/argon2).

## Contact

Feel free to ask questions using these contact details:

**Current Maintainers:**

* Josh Buker ([@joshbuker](https://github.com/joshbuker)) | [Email](mailto:crypto+sorcery@joshbuker.com?subject=Sorcery)

## License

This gem is available as open source under the terms of the
[MIT License](https://opensource.org/licenses/MIT).


CHANGELOG.md
# Changelog

Historical changelog for all versions.

## HEAD

## v1.2.0

* Synced with latest upstream changes (technion/ruby-argon2 `v2.2.0`)

## v1.1.0

* Added support for passing parallelism cost to `Argon2::Password.create`

## v1.0.0

This project has been forked from
[Technion's original argon2 wrapper](https://github.com/technion/ruby-argon2).

If you previously used `argon2` and would like to update to `sorcery-argon2`,
please see: [Migrating from `argon2` to `sorcery-argon2`](README.md#migrating-from-argon2-to-sorcery-argon2)

Changes between `argon2 - 2.0.3` and `sorcery-argon2 - 1.0.0`:

* Refactored Argon2::Password to include additional helpers and simplify hash
  creation.
* Renamed top level exception from: `Argon2::ArgonHashHail` to: `Argon2::Error`
* Added new exceptions that inherit from the top level exception.


test/api_test.rb
# frozen_string_literal: true

require 'test_helper'

class Argon2APITest < Minitest::Test
  def test_create_default
    assert pass = Argon2::Password.create('mypassword')
    assert_instance_of Argon2::Password, pass
    assert_equal 16, pass.m_cost
    assert_equal 2, pass.t_cost
    assert_equal 1, pass.p_cost
    # assert_nil pass.secret # Secret is not persisted on the Argon2::Password
  end

  def test_create_args
    assert pass = Argon2::Password.create('mypassword', t_cost: 4, m_cost: 12, p_cost: 4)
    assert_instance_of Argon2::Password, pass
    assert_equal 12, pass.m_cost
    assert_equal 4, pass.t_cost
    assert_equal 4, pass.p_cost
    # assert_nil pass.secret # Secret is not persisted on the Argon2::Password
  end

  # For usage of the secret param, see key_test.rb
  def test_secret
    assert pass = Argon2::Password.create('mypassword', secret: "A secret")
    skip("The secret isn't kept on the Argon2::Password instance")
    assert_equal pass.secret, "A secret"
  end

  def test_hash
    assert Argon2::Password.create('mypassword')
  end

  def test_valid_hash
    secure_pass = Argon2::Password.create('A secret')
    assert Argon2::Password.valid_hash?(secure_pass)
  end

  ###############
  ## New Tests ##
  ###############
  ORIGINAL_PASSWORD = 'mypassword'
  PEPPER = 'A secret'

  def test_create_password_without_parameters
    assert argon2 = Argon2::Password.create(ORIGINAL_PASSWORD)

    assert argon2.is_a?(Argon2::Password)
    assert_equal argon2.m_cost, 16
    assert_equal argon2.t_cost, 2
    assert argon2.matches?(ORIGINAL_PASSWORD)
    assert Argon2::Password.verify_password(ORIGINAL_PASSWORD, argon2)
  end

  def test_create_password_with_parameters
    assert argon2 = Argon2::Password.create(ORIGINAL_PASSWORD, t_cost: 4, m_cost: 12)

    assert argon2.is_a?(Argon2::Password)
    assert_equal argon2.m_cost, 12
    assert_equal argon2.t_cost, 4
    assert argon2.matches?(ORIGINAL_PASSWORD)
    assert Argon2::Password.verify_password(ORIGINAL_PASSWORD, argon2)
  end

  def test_create_password_with_secret
    assert argon2 = Argon2::Password.create(ORIGINAL_PASSWORD, secret: PEPPER)

    assert argon2.is_a?(Argon2::Password)
    assert_equal argon2.m_cost, 16
    assert_equal argon2.t_cost, 2
    assert argon2.matches?(ORIGINAL_PASSWORD, PEPPER)
    assert Argon2::Password.verify_password(ORIGINAL_PASSWORD, argon2, PEPPER)
  end
end


test/error_test.rb
# frozen_string_literal: true

require 'test_helper'

class Argon2ErrorTest < Minitest::Test
  def test_ffi_fail
    assert_raises Argon2::Error do
      Argon2::Engine.hash_argon2i("password", "somesalt\0\0\0\0\0\0\0\0", 2, 1, 1)
    end
  end

  def test_memory_too_small
    assert_raises Argon2::Error do
      Argon2::Engine.hash_argon2id_encode("password",
                                          "somesalt\0\0\0\0\0\0\0\0", 2, 1, 1, nil)
    end
  end

  def test_salt_size
    assert_raises Argon2::Error do
      Argon2::Engine.hash_argon2id_encode("password", "somesalt", 2, 16, 1, nil)
    end
  end

  def test_passwd_null
    assert_raises Argon2::Error do
      Argon2::Engine.hash_argon2id_encode(nil, "somesalt\0\0\0\0\0\0\0\0", 2,
                                          16, 1, nil)
    end
  end

  def test_error_inheritance
    assert Argon2::Error                       < StandardError
    assert Argon2::Errors::InvalidHash         < Argon2::Error
    assert Argon2::Errors::InvalidVersion      < Argon2::Errors::InvalidHash
    assert Argon2::Errors::InvalidCost         < Argon2::Errors::InvalidHash
    assert Argon2::Errors::InvalidTCost        < Argon2::Errors::InvalidCost
    assert Argon2::Errors::InvalidMCost        < Argon2::Errors::InvalidCost
    assert Argon2::Errors::InvalidPCost        < Argon2::Errors::InvalidCost
    assert Argon2::Errors::InvalidPassword     < Argon2::Error
    assert Argon2::Errors::InvalidSaltSize     < Argon2::Error
    assert Argon2::Errors::InvalidOutputLength < Argon2::Error
    assert Argon2::Errors::ExtError            < Argon2::Error
  end

  def test_invalid_password_msg
    assert_equal Argon2::Errors::InvalidPassword.new.message,
                 "Invalid password (expected a String)"
  end
end


ext/argon2_wrap/test.c
/* Wrapper for argon Ruby bindings
 * lolware.net
 * Much of this code is based on run.c from the reference implementation
 */
#include <stdio.h>
#include <stdint.h>
#include <inttypes.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <assert.h>

#include "argon2.h"

#define OUT_LEN 32
#define SALT_LEN 16

/**
 * Hashes a password with Argon2i, producing a raw hash
 * @param t_cost Number of iterations
 * @param m_cost Sets memory usage to 2^m_cost kibibytes
 * @param parallelism Number of threads and compute lanes
 * @param pwd Pointer to password
 * @param pwdlen Password size in bytes
 * @param salt Pointer to salt
 * @param saltlen Salt size in bytes
 * @param hash Buffer where to write the raw hash
 * @param hashlen Desired length of the hash in bytes
 * @pre   Different parallelism levels will give different results
 * @pre   Returns ARGON2_OK if successful

int argon2i_hash_raw(const uint32_t t_cost, const uint32_t m_cost,
                     const uint32_t parallelism, const void *pwd,
                     const size_t pwdlen, const void *salt,
                     const size_t saltlen, void *hash, const size_t hashlen);

*/

unsigned int argon2_wrap(char *out, const char *pwd, size_t pwd_length,
    uint8_t *salt, uint32_t saltlen, uint32_t t_cost, uint32_t m_cost,
    uint32_t lanes, uint8_t *secret, size_t secretlen);

int wrap_argon2_verify(const char *encoded, const char *pwd,
    const size_t pwdlen,
    uint8_t *secret, size_t secretlen);


int main()
{
    unsigned char out[OUT_LEN];
    unsigned char hex_out[OUT_LEN*2 + 4]; /* Allow space for NULL byute */
    char out2[300];
    char *pwd = NULL;
    uint8_t salt[SALT_LEN];
    int i, ret;

    memset(salt, 0x00, SALT_LEN); /* pad with null bytes */
    memcpy(salt, "somesalt", 8);


#define RAWTEST(T, M, P, PWD, REF) \
    pwd = strdup(PWD); \
    assert(pwd); \
    ret = argon2i_hash_raw(T, 1<<M, P, pwd, strlen(pwd), salt, SALT_LEN, out, OUT_LEN); \
    assert(ret == ARGON2_OK); \
    for(i=0; i<OUT_LEN; ++i ) \
        sprintf((char*)(hex_out + i*2), "%02x", out[i] ); \
    assert(memcmp(hex_out, REF, OUT_LEN*2) == 0); \
    free(pwd); \
    printf( "Ref test: %s: PASS\n", REF);

    RAWTEST(2, 16, 1, "password", "1c7eeef9e0e969b3024722fc864a1ca9f6ca20da73f9bf3f1731881beae2039e");
    RAWTEST(2, 18, 1, "password", "5c6dfd2712110cf88f1426059b01d87f8210d5368da0e7ee68586e9d4af4954b");
    RAWTEST(2, 8, 1, "password", "dfebf9d4eadd6859f4cc6a9bb20043fd9da7e1e36bdacdbb05ca569f463269f8");
    RAWTEST(1, 16, 1, "password", "fabd1ddbd86a101d326ac2abe79660202b10192925d2fd2483085df94df0c91a");
    RAWTEST(4, 16, 1, "password", "b3b4cb3d6e2c1cb1e7bffdb966ab3ceafae701d6b7789c3f1e6c6b22d82d99d5");
    RAWTEST(2, 16, 1, "differentpassword", "b2db9d7c0d1288951aec4b6e1cd3835ea29a7da2ac13e6f48554a26b127146f9");
    memcpy(salt, "diffsalt", 8);
    RAWTEST(2, 16, 1, "password", "bb6686865f2c1093f70f543c9535f807d5b42d5dc6d71f14a4a7a291913e05e0");


#define WRAP_TEST(T, M, PWD, REF) \
    pwd = strdup(PWD); \
    argon2_wrap(out2, pwd, strlen(PWD), salt, strlen((const char *)salt),T, 1<<M, 1, NULL, 0); \
    free(pwd); \
    assert(memcmp(out2, REF, strlen(REF)) == 0); \
    printf( "Ref test: %s: PASS\n", REF);

    memcpy(salt, "somesalt", 8);
    WRAP_TEST(2, 16, "password",
            "$argon2id$v=19$m=65536,t=2,p=1$c29tZXNhbHQ$CTFhFdXPJO1aFaMaO6Mm5c8y7cJHAph8ArZWb2GRPPc");

    WRAP_TEST(2, 8, "password",
            "$argon2id$v=19$m=256,t=2,p=1$c29tZXNhbHQ$nf65EOgLrQMR/uIPnA4rEsF5h7TKyQwu9U1bMCHGi/4");

    WRAP_TEST(2, 16, "differentpassword",
            "$argon2id$v=19$m=65536,t=2,p=1$c29tZXNhbHQ$C4TWUs9rDEvq7w3+J4umqA32aWKB1+DSiRuBfYxFj94");

    ret = wrap_argon2_verify("$argon2i$v=19$m=256,t=2,p=1$c29tZXNhbHQAAAAAAAAAAA$3+v51OrdaFn0zGqbsgBD/Z2n4eNr2s27BcpWn0Yyafg", "password",
            strlen("password"), NULL, 0);
    assert(ret == ARGON2_OK);
    printf("Verify OK test i variant: PASS\n");

    ret = wrap_argon2_verify("$argon2id$v=19$m=65536,t=2,p=1$c29tZXNhbHQ$CTFhFdXPJO1aFaMaO6Mm5c8y7cJHAph8ArZWb2GRPPc", "password",
            strlen("password"), NULL, 0);
    assert(ret == ARGON2_OK);
    printf("Verify OK test id variant: PASS\n");

    ret = wrap_argon2_verify("$argon2i$v=19$m=65536,t=2,p=1$c29tZXNhbHQAAAAAAAAAAA$iUr0/y4tJvPOFfd6fhwl20W04gQ56ZYXcroZnK3bAB4", "notpassword",
            strlen("notpassword"), NULL, 0);
    assert(ret == ARGON2_DECODING_FAIL);
    printf("Verify FAIL test: PASS\n");
    return 0;


}



lib/argon2/password.rb
# frozen_string_literal: true

module Argon2
  ##
  # Front-end API for the Argon2 module.
  #
  class Password
    # Used as the default time cost if one isn't provided when calling
    # Argon2::Password.create
    DEFAULT_T_COST = 2
    # Used to validate the minimum acceptable time cost
    MIN_T_COST = 1
    # Used to validate the maximum acceptable time cost
    MAX_T_COST = 750
    # Used as the default memory cost if one isn't provided when calling
    # Argon2::Password.create
    DEFAULT_M_COST = 16
    # Used to validate the minimum acceptable memory cost
    MIN_M_COST = 3
    # Used to validate the maximum acceptable memory cost
    MAX_M_COST = 31
    # Used as the default parallelism cost if one isn't provided when calling
    # Argon2::Password.create
    DEFAULT_P_COST = 1
    # Used to validate the minimum acceptable parallelism cost
    MIN_P_COST = 1
    # Used to validate the maximum acceptable parallelism cost
    MAX_P_COST = 8
    # The complete Argon2 digest string (not to be confused with the checksum).
    attr_reader :digest
    # The hash portion of the stored password hash.
    attr_reader :checksum
    # The salt of the stored password hash.
    attr_reader :salt
    # Variant used (argon2i / argon2d / argon2id)
    attr_reader :variant
    # The version of the argon2 algorithm used to create the hash.
    attr_reader :version
    # The time cost factor used to create the hash.
    attr_reader :t_cost
    # The memory cost factor used to create the hash.
    attr_reader :m_cost
    # The parallelism cost factor used to create the hash.
    attr_reader :p_cost

    ##
    # Class methods
    #
    class << self
      ##
      # Takes a user provided password and returns an Argon2::Password instance
      # with the resulting Argon2 hash.
      #
      # Usage:
      #
      #    Argon2::Password.create(password)
      #    Argon2::Password.create(password, t_cost: 4, m_cost: 20)
      #    Argon2::Password.create(password, secret: pepper)
      #    Argon2::Password.create(password, m_cost: 17, secret: pepper)
      #
      # Currently available options:
      #
      # * :t_cost
      # * :m_cost
      # * :p_cost
      # * :secret
      #
      def create(password, options = {})
        raise Argon2::Errors::InvalidPassword unless password.is_a?(String)

        t_cost = options[:t_cost] || DEFAULT_T_COST
        m_cost = options[:m_cost] || DEFAULT_M_COST
        p_cost = options[:p_cost] || DEFAULT_P_COST

        raise Argon2::Errors::InvalidTCost if t_cost < MIN_T_COST || t_cost > MAX_T_COST
        raise Argon2::Errors::InvalidMCost if m_cost < MIN_M_COST || m_cost > MAX_M_COST
        raise Argon2::Errors::InvalidPCost if p_cost < MIN_P_COST || p_cost > MAX_P_COST

        salt = Engine.saltgen
        secret = options[:secret]

        Argon2::Password.new(
          Argon2::Engine.hash_argon2id_encode(
            password, salt, t_cost, m_cost, p_cost, secret
          )
        )
      end

      ##
      # Regex to validate if the provided String is a valid Argon2 hash output.
      #
      # Supports 1 and argon2id formats.
      #
      def valid_hash?(digest)
        /^\$argon2(id?|d).{,113}/ =~ digest
      end

      ##
      # Takes a password, Argon2 hash, and optionally a secret, then uses the
      # Argon2 C Library to verify if they match.
      #
      # Also accepts passing another Argon2::Password instance as the password,
      # in which case it will compare the final Argon2 hash for each against
      # each other.
      #
      # Usage:
      #
      #    Argon2::Password.verify_password(password, argon2_hash)
      #    Argon2::Password.verify_password(password, argon2_hash, secret)
      #
      def verify_password(password, digest, secret = nil)
        digest = digest.to_s
        if password.is_a?(Argon2::Password)
          password == Argon2::Password.new(digest)
        else
          Argon2::Engine.argon2_verify(password, digest, secret)
        end
      end
    end

    ######################
    ## Instance Methods ##
    ######################

    ##
    # Initialize an Argon2::Password instance using any valid Argon2 digest.
    #
    def initialize(digest)
      digest = digest.to_s

      raise Argon2::Errors::InvalidHash unless valid_hash?(digest)

      # Split the digest into its component pieces
      split_digest = split_hash(digest)
      # Assign each piece to the Argon2::Password instance
      @digest   = digest
      @variant  = split_digest[:variant]
      @version  = split_digest[:version]
      @t_cost   = split_digest[:t_cost]
      @m_cost   = split_digest[:m_cost]
      @p_cost   = split_digest[:p_cost]
      @salt     = split_digest[:salt]
      @checksum = split_digest[:checksum]
    end

    ##
    # Helper function to allow easily comparing an Argon2::Password against the
    # provided password and secret.
    #
    def matches?(password, secret = nil)
      self.class.verify_password(password, digest, secret)
    end

    ##
    # Compares two Argon2::Password instances to see if they come from the same
    # digest/hash.
    #
    def ==(other)
      # TODO: Should this return false instead of raising an error?
      unless other.is_a?(Argon2::Password)
        raise ArgumentError,
              'Can only compare an Argon2::Password against another Argon2::Password'
      end

      digest == other.digest
    end

    ##
    # Converts an Argon2::Password instance into a String.
    #
    def to_s
      digest.to_s
    end

    ##
    # Converts an Argon2::Password instance into a String.
    #
    def to_str
      digest.to_str
    end

    private

    ##
    # Helper method to allow checking if a hash is valid in the initializer.
    #
    def valid_hash?(digest)
      self.class.valid_hash?(digest)
    end

    # FIXME: Reduce complexity/AbcSize
    # rubocop:disable Metrics/AbcSize

    ##
    # Helper method to extract the various values from a digest into attributes.
    #
    def split_hash(digest)
      # TODO: Is there a better way to explode the digest into attributes?
      _, variant, version, config, salt, checksum = digest.split('$')
      # Regex magic to extract the values for each setting
      version = /v=(\d+)/.match(version)
      t_cost  = /t=(\d+),/.match(config)
      m_cost  = /m=(\d+),/.match(config)
      p_cost  = /p=(\d+)/.match(config)

      # Make sure none of the values are missing
      raise Argon2::Errors::InvalidVersion if version.nil?
      raise Argon2::Errors::InvalidTCost   if t_cost.nil?
      raise Argon2::Errors::InvalidMCost   if m_cost.nil?
      raise Argon2::Errors::InvalidPCost   if p_cost.nil?

      # Undo the 2^m_cost operation when encoding the hash to get the original
      # m_cost input back.
      m_cost = Math.log2(m_cost[1].to_i).to_i

      {
        variant: variant.to_str,
        version: version[1].to_i,
        t_cost: t_cost[1].to_i,
        m_cost: m_cost,
        p_cost: p_cost[1].to_i,
        salt: salt.to_str,
        checksum: checksum.to_str
      }
    end
    # rubocop:enable Metrics/AbcSize
  end
end


ext/argon2_wrap/argon_wrap.c
/* Wrapper for argon Ruby bindings
 * lolware.net
 * Much of this code is based on run.c from the reference implementation
 */

#include <stdio.h>
#include <stdint.h>
#include <inttypes.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#include "argon2.h"
#include "core.h"
#include "encoding.h"

#define T_COST_DEF 3
#define LOG_M_COST_DEF 12 /* 2^12 = 4 MiB */
#define LANES_DEF 1
#define THREADS_DEF 1
#define OUT_LEN 32
#define SALT_LEN 16
#define ENCODE_LEN 96 /* Does not include SALT LEN */

/* Workaround for https://github.com/technion/ruby-argon2/issues/8. Hopefully temporary */
static int wrap_compare(const uint8_t *b1, const uint8_t *b2, size_t len) {
    size_t i;
    uint8_t d = 0U;

    for (i = 0U; i < len; i++) {
        d |= b1[i] ^ b2[i];
    }
    return (int)((1 & ((d - 1) >> 8)) - 1);
}

int argon2_wrap_version(char *out, const char *pwd, size_t pwd_length,
        uint8_t *salt,  uint32_t saltlen, uint32_t t_cost, uint32_t m_cost,
        uint32_t lanes, uint8_t *secret, size_t secretlen, uint32_t version,
        argon2_type type)
{
    uint8_t hash[OUT_LEN];
    argon2_context context;

    if (!pwd) {
        return ARGON2_PWD_PTR_MISMATCH;
    }

    if (!salt) {
        return ARGON2_PWD_PTR_MISMATCH;
    }

    context.out = hash;
    context.outlen = OUT_LEN;
    context.pwd = (uint8_t *)pwd;
    context.pwdlen = pwd_length;
    context.salt = salt;
    context.saltlen = saltlen;
    context.secret = secret;
    context.secretlen = secretlen;
    context.ad = NULL;
    context.adlen = 0;
    context.t_cost = t_cost;
    context.m_cost = m_cost;
    context.lanes = lanes;
    context.threads = lanes;
    context.allocate_cbk = NULL;
    context.free_cbk = NULL;
    context.flags = 0;
    context.version = version;

    int result;
    if (type == Argon2_i) {
        result = argon2i_ctx(&context);
    }  else if (type == Argon2_id) {
        result = argon2id_ctx(&context);
    } else if (type == Argon2_d) {
        result = argon2d_ctx(&context);
    } else {
        // Unsupported type
        return ARGON2_ENCODING_FAIL;
    }

    if (result != ARGON2_OK)
        return result;

    encode_string(out, ENCODE_LEN + saltlen, &context, type);
    return ARGON2_OK;
}

/* Since all new hashes will use latest version, this wraps the
   * function including the version
   */
int argon2_wrap(char *out, const char *pwd, size_t pwd_length,
        uint8_t *salt,  uint32_t saltlen, uint32_t t_cost, uint32_t m_cost,
        uint32_t lanes, uint8_t *secret, size_t secretlen)
{
    return argon2_wrap_version(out, pwd, pwd_length, salt, saltlen,
            t_cost, m_cost, lanes, secret, secretlen, ARGON2_VERSION_13, Argon2_id);
}

int wrap_argon2_verify(const char *encoded, const char *pwd,
    const size_t pwdlen,
    uint8_t *secret, size_t secretlen)
{
    argon2_context ctx;
    int ret;
    char *out;
    memset(&ctx, 0, sizeof(argon2_context));
    size_t encoded_len;
    argon2_type type;

    encoded_len = strlen(encoded);
    /* larger than max possible values */
    ctx.saltlen = encoded_len;
    ctx.outlen = encoded_len;

    ctx.salt = malloc(ctx.saltlen);
    ctx.out = malloc(ctx.outlen);
    if (!ctx.out || !ctx.salt) {
        free(ctx.salt);
        free(ctx.out);
        return ARGON2_MEMORY_ALLOCATION_ERROR;
    }

    if (memcmp(encoded, "$argon2id", strlen("$argon2id")) == 0) {
        type = Argon2_id;
    } else if (memcmp(encoded, "$argon2i", strlen("$argon2i")) == 0) {
        type = Argon2_i;
    } else if (memcmp(encoded, "$argon2d", strlen("$argon2d")) == 0) {
        type = Argon2_d;
    } else {
        // Other types not yet supported
        free(ctx.salt);
        free(ctx.out);
        return ARGON2_DECODING_FAIL;
    }

    if (decode_string(&ctx, encoded, type) != ARGON2_OK) {
        free(ctx.salt);
        free(ctx.out);
        return ARGON2_DECODING_FAIL;
    }

    out = malloc(ENCODE_LEN + ctx.saltlen);
    if(!out) {
        free(ctx.salt);
        free(ctx.out);
        return ARGON2_DECODING_FAIL;
    }

    ret = argon2_wrap_version(out, pwd, pwdlen, ctx.salt, ctx.saltlen,
            ctx.t_cost, ctx.m_cost, ctx.lanes, secret, secretlen,
            ctx.version, type);

    free(ctx.salt);

    if (ret != ARGON2_OK || wrap_compare((uint8_t*)out, (uint8_t*)encoded,
                strlen(encoded))) {
        free(ctx.out);
        free(out);
        return ARGON2_DECODING_FAIL;
    }
    free(ctx.out);
    free(out);

    return ARGON2_OK;
}
